odl/trafos/wavelet.py

```diff
@@ -136,7 +244,112 @@ def coeff_size_list(shape, nscales, wbasis, mode):
     return size_list
 
 
-def pywt_coeff_to_array(coeff, size_list):
+def pywt_dict_to_array(coeffs, size_list):
+    """Convert a PyWavelet coefficiet dictionary into a flat array.
+
+    Related to 1D, 2D and 3D discrete wavelet transforms with `axes` option.
+
+    Parameters
+    ----------
+    coeff : ordered `dict`
+        Coefficients are organized in the dictionary with the following keys:
+
+        In 1D:
+        ``a`` = approximation,
+
+        ``d`` = detail
+
+        In 2D:
+
+        ``aa`` = approx. on 1st dim, approx. on 2nd dim (approximation),
+
+        ``ad`` = approx. on 1st dim, detail on 2nd dim (horizontal),
+
+        ``da`` = detail on 1st dim, approx. on 2nd dim (vertical),
+
+        ``dd`` = detail on 1st dim, detail on 2nd dim (diagonal),
+
+        In 3D:
+        ``aaa`` = approx. on 1st dim, approx. on 2nd dim, approx. on 3rd dim,
+
+        ``aad`` = approx. on 1st dim, approx. on 2nd dim, detail on 3rd dim,
+
+        ``ada`` = approx. on 1st dim, detail on 3nd dim, approx. on 3rd dim,
+
+        ``add`` = approx. on 1st dim, detail on 3nd dim, detail on 3rd dim,
+
+        ``daa`` = detail on 1st dim, approx. on 2nd dim, approx. on 3rd dim,
+
+        ``dad`` = detail on 1st dim, approx. on 2nd dim, detail on 3rd dim,
+
+        ``dda`` = detail on 1st dim, detail on 2nd dim, approx. on 3rd dim,
+
+        ``ddd`` = detail on 1st dim, detail on 2nd dim, detail on 3rd dim,
+
+    size_list : `list`
+        A list containing the sizes of the wavelet (approximation
+        and detail) coefficients when `axes` option is used.
+
+    Returns
+    -------
+    arr : `numpy.ndarray`
+        Flattened and concatenated coefficient array
+        The length of the array depends on the size of input image to
+        be transformed and on the chosen wavelet basis.
+    """
+    flat_sizes = [np.prod(shp) for shp in size_list[:-1]]
+    ndim = len(size_list[0])
+    dcoeffs_per_scale = 2 ** ndim - 1
+
+    flat_total_size = flat_sizes[0] + dcoeffs_per_scale * sum(flat_sizes[1:])
+    flat_coeff = np.empty(flat_total_size)
+
+    start = 0
+    stop = flat_sizes[0]
+
+    if ndim == 1:
+        a = coeffs['a']
+        d = coeffs['d']
+        flat_coeff[start:stop] = a.ravel()
+        start, stop = stop, stop + flat_sizes[1]
+        flat_coeff[start:stop] = d.ravel()
+
+    elif ndim == 2:
+        aa = coeffs['aa']
+        ad = coeffs['ad']
+        da = coeffs['da']
+        dd = coeffs['dd']
+        details = (ad, da, dd)
```

Comment:
Further improved the integration of `wavelet.py` with `PyWavelet`:

1.  use new PyWavelet functions `waverecn` and `wavedecn` to compute the multiresolution wavelet transform in 3D

2. Related to 1. modified the helper functions and tests

Also some of the review comments are used in this version.